src/Features/CSharp/Portable/CodeCleanup/CSharpCodeCleanupService.cs

```diff
@@ -107,44 +108,69 @@ internal class CSharpCodeCleanupService : ICodeCleanupService
             );
         }
 
-        public async Task<Document> CleanupAsync(Document document, CancellationToken cancellationToken)
+        public async Task<Document> CleanupAsync(
+            Document document, IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
             var docOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
 
+            // add one item for the 'Remove Usings', one for 'Sort using' and one item for the
+            // 'format' actions we'll do last
+            progressTracker.AddItems(3);
+
             if (_codeFixServiceOpt != null)
             {
-                document = await ApplyCodeFixesAsync(document, docOptions, cancellationToken).ConfigureAwait(false);
+                document = await ApplyCodeFixesAsync(
+                    document, progressTracker, docOptions, cancellationToken).ConfigureAwait(false);
             }
 
             // do the remove usings after code fix, as code fix might remove some code which can results in unused usings.
-            document = await RemoveSortUsingsAsync(document, docOptions, cancellationToken).ConfigureAwait(false);
+            document = await RemoveSortUsingsAsync(
+                document, progressTracker, docOptions, cancellationToken).ConfigureAwait(false);
+
+            progressTracker.Description = FeaturesResources.Formatting_document;
+            var result = await Formatter.FormatAsync(document).ConfigureAwait(false);
+            progressTracker.ItemCompleted();
 
-            return await Formatter.FormatAsync(document).ConfigureAwait(false);
+            return result;
         }
 
-        private async Task<Document> RemoveSortUsingsAsync(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        private async Task<Document> RemoveSortUsingsAsync(
+            Document document, IProgressTracker tracker,  DocumentOptionSet docOptions, CancellationToken cancellationToken)
```

Comment:
Note: IProgressTracker is actually a superior API for IDE features over IProgress for many situations.  Specifically, it encapsulates the concurrency issues inherent with trying to track completed/total across a heavily concurrent algorithm.

With IProgressTracker algorithms (like FindRefs/NavTo/etc.) only need to say "here's how many work-items i will perform" and then only need to say when they've completed a "work item".    This can all happen in parallel, and the values and reporting will all be managed in a threadsafe manner.